egui_demo_lib/src/apps/demo/zoom_rotate.rs

```diff
@@ -0,0 +1,117 @@
+            let to_screen = RectTransform::from_to(
+                Rect::from_min_size(Pos2::ZERO - painter_proportions, 2. * painter_proportions),
+                response.rect,
+            );
+
+            // check for touch input (or the lack thereof) and update zoom and scale factors, plus
+            // color and width:
+            let mut stroke_width = 1.;
+            let mut color = Color32::GRAY;
+            if let Some(multi_touch) = ui.input().multi_touch() {
+                // This adjusts the current zoom factor and rotation angle according to the dynamic
+                // change (for the current frame) of the touch gesture:
+                self.zoom *= multi_touch.zoom_delta;
+                self.rotation += multi_touch.rotation_delta;
+                // touch pressure shall make the arrow thicker (not all touch devices support this):
+                stroke_width += 10. * multi_touch.force;
+                // the drawing color depends on the number of touches:
+                color = match multi_touch.num_touches {
+                    2 => Color32::GREEN,
+                    3 => Color32::BLUE,
+                    4 => Color32::YELLOW,
+                    _ => Color32::RED,
+                };
+                // for a smooth touch experience (not strictly required, but I had the impression
+                // that it helps to reduce some lag, especially for the initial touch):
+                ui.ctx().request_repaint();
+            } else if let Some(last_time) = self.time_of_last_update {
+                // This has nothing to do with the touch gesture. It just smoothly brings the
+                // painted arrow back into its original position, for a nice visual effect:
+                let dt = ui.input().time - last_time;
+                const ZOOM_ROTATE_HALF_LIFE: f64 = 1.; // time[sec] after which half the amount of zoom/rotation will be reverted
+                let half_life_factor = (-(2_f64.ln()) / ZOOM_ROTATE_HALF_LIFE * dt).exp() as f32;
+                self.zoom = 1. + ((self.zoom - 1.) * half_life_factor);
+                self.rotation *= half_life_factor;
+                // this is an animation, so we want real-time UI updates:
+                ui.ctx().request_repaint();
+            }
+
+            let zoom_and_rotate = self.zoom * Rot2::from_angle(self.rotation);
+
+            // Paints an arrow pointing from bottom-left (-0.5, 0.5) to top-right (0.5, -0.5),
+            // but scaled and rotated according to the current translation:
+            let arrow_start = zoom_and_rotate * vec2(-0.5, 0.5);
+            let arrow_direction = zoom_and_rotate * vec2(1., -1.);
+            painter.arrow(
+                to_screen * (Pos2::ZERO + arrow_start),
+                to_screen.scale() * arrow_direction,
+                Stroke::new(stroke_width, color),
+            );
+            self.time_of_last_update = Some(ui.input().time);
+        });
```

Comment:
It would be great if this demo also showed translation; then it would demonstrate all the higher level parts of the multitouch gesture API

Reply:
OK, I will add this. Maybe tomorrow…

I was a bit reluctant to add translation because, for some devices (like a touch pad), the units for translation would neither be points, nor pixels on the screen, but rather the physical resolution of the touch device itself. Therefore, it feels like 'cheating' to use the translation coordinates just as if they were real screen coordinates – although it would work just fine on touch *screens*.

Unfortunately, I am only able to test all this with an iPhone and iPad. Both have (obviously) touch *screens*, not touch *pads*, so the resolution of the touch device and screen are always identical. I have a track pad, but `winit` does not support touch events for it. Same for `web_sys` in any browser I have tried.

Reply:
Resolved with <https://github.com/emilk/egui/pull/306/commits/755ca301a08b3d99fc0cca9dca9c7642c4be5d16>.
The [online demo](https://quadruple-output.github.io/egui/) is up-to-date.